### PR TITLE
subsys/testsuite/ztest: create stub headers at CMake configure time

### PR DESCRIPTION
**Pull request by ycsin**
These header stubs seems to be required by the `unit_testing` board only. Instead of having them in the source, let's remove them and only generate them at CMake configure time.

~~See also #69442, which relocates these headers instead of having them generated at build~~

### CODE REVIEW

File: subsys/testsuite/ztest/include/zephyr/ztest.h [hunk 1]
```diff
@@ -22,19 +22,6 @@
...
-#ifdef __cplusplus
-extern "C" {
-#endif
-struct esf;
-typedef struct esf z_arch_esf_t;
-#ifdef __cplusplus
-}
-#endif
-#endif /* KERNEL */
-
```
Comment (ycsin) [inline subsys/testsuite/ztest/include/zephyr/ztest.h:37]: removed as they don't seem required by upstream CI, can be split into another commit

**Review by ycsin — COMMENTED** [review]
ping

File: cmake/modules/unittest.cmake [hunk 2]
```diff
@@ -55,6 +55,15 @@ include(${ZEPHYR_BASE}/cmake/kobj.cmake)
 add_dependencies(test_interface ${KOBJ_TYPES_H_TARGET})
 gen_kobj(KOBJ_GEN_DIR)
 
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/offsets.h)
```
Comment (tejlmand) [inline cmake/modules/unittest.cmake:58]: Minor comment related to commit message.

This is not done at build time, but at CMake configure time.

File: cmake/modules/unittest.cmake [hunk 3]
```diff
@@ -55,6 +55,15 @@ include(${ZEPHYR_BASE}/cmake/kobj.cmake)
...
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/syscall_list.h)
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/devicetree_generated.h)
```
Comment (tejlmand) [inline cmake/modules/unittest.cmake:60]: other CMake function calls are written in lower case, so please adjust to `file(...)` if you continue this approach.

File: include/zephyr/linker/linker-defs.h [hunk 4]
```diff
@@ -25,14 +25,7 @@
 #include <zephyr/sys/util.h>
 #include <offsets.h>
 
-/* We need to dummy out DT_NODE_HAS_STATUS when building the unittests.
- * Including devicetree.h would require generating dummy header files
- * to match what gen_defines creates, so it's easier to just dummy out
- * DT_NODE_HAS_STATUS.
- */
-#ifdef ZTEST_UNITTEST
-#define DT_NODE_HAS_STATUS(node, status) 0
```
Comment (tejlmand) [inline include/zephyr/linker/linker-defs.h:34]: Curious, where is `DT_NODE_HAS_STATUS(node, status)` defined now / why is it no longer needed ?

File: cmake/modules/unittest.cmake (hunk 2 above)
Comment (ycsin) [reply cmake/modules/unittest.cmake]: Fixed, thanks

File: cmake/modules/unittest.cmake (hunk 3 above)
Comment (ycsin) [reply cmake/modules/unittest.cmake]: How did I miss that? Thanks

File: include/zephyr/linker/linker-defs.h (hunk 4 above)
Comment (ycsin) [reply include/zephyr/linker/linker-defs.h]: This change is probably not required, removed.

**Review by cfriedt — DISMISSED** [review]

File: cmake/modules/unittest.cmake [hunk 5]
```diff
@@ -58,8 +58,22 @@ include(${ZEPHYR_BASE}/cmake/kobj.cmake)
 add_dependencies(test_interface ${KOBJ_TYPES_H_TARGET})
 gen_kobj(KOBJ_GEN_DIR)
 
+# Generates empty header files to build
+file(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/devicetree_generated.h)
```
Comment (ycsin) [inline cmake/modules/unittest.cmake:62]: is it more elegant to do a list + foreach instead?

Comment (nashif) [reply cmake/modules/unittest.cmake:62]: yes, I guess.

**Review by nashif — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

File: cmake/modules/unittest.cmake (hunk 5 above)
Comment (ycsin) [reply cmake/modules/unittest.cmake]: converted to list + foreach

**Review by cfriedt — APPROVED** [review]

**Comment by ycsin** [comment]
ping - @nashif could you PTAL again? Thanks!

**Comment by ycsin** [comment]
another ping @nashif

**Review by nashif — APPROVED** [review]

**Pull request merged**